### DISCUSSION TITLE
Validate joins match rows as often as expected

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#10819] [Ideas]
It'd be super convenient to be able to assert that rows appear in the output of a joins as often as one expects. 

I'm very unsure about the syntax, but I imagine something like 
```
SELECT
*
FROM
df1
JOIN ASSERT (<=1, )
df2
ON
condition
```
Could throw if any row from `df1` appears more than once in the output (i.e. was matched multiple times, not appears twice just because the left table has duplicate rows) whereas `JOIN ASSERT (, <=1)` would swap the tables in the assertion and  `JOIN (<=1, <=1)` would do both. 

Similarly, `>=1` / `=1` (or even more exotic `>3`?) could assert the obvious corresponding statements.

[Polars ](https://docs.pola.rs/py-polars/html/reference/dataframe/api/polars.DataFrame.join.html) calls the `<=1` asserts many-to-one/one-to-many/one-to-one but I would find a `JOIN ONE-TO-MANY` syntax slightly worse than what I proposed above, first because it suggests non-zero and larger-than-one cardinality checks which aren't performed, second because it just took me a while to deduce which way around it maps to my proposal, third because it doesn't easily generalize to `>=1` / `=1`, and fourth because one might think it changes semantics of the join, e.g, by dropping rows or adding null rows, as opposed to being purely an assertion. 

(One might also think about `JOIN ENFORCE (x, y)` to generalize outer joins, but that's a different story that's probably more complicated.)

**Comment by Tishj** [comment]
Maybe that could work as an EXPLAIN ANALYZE metric

**Comment by soerenwolfers** [comment]
duplicate of https://github.com/duckdb/duckdb/discussions/10615
